### ISSUE TITLE
AOT fails at runtime trying to execute an extern method with error:  "System.BadImageFormatException: Method has no body".

### ISSUE THREAD

**Issue by BradChase2011** [unoplatform/Uno.Wasm.Bootstrap#885]
I have a bunch of dynamic classes that run interop code in the browser and everything runs perfectly well until I try to AOT.  I am using "InterpreterAndAOT".  That said, here is one of the dynamic classes:

```
public class MainBindingGroupNoReturn<_A1, _A2, _A3, _A4, _A5, _A6> : MainBindingGroup
{
    public Action<_A1, _A2, _A3, _A4, _A5, _A6> Callback { get; set; }
    public _A1 Arg1 { get; set; }
    public _A2 Arg2 { get; set; }
    public _A3 Arg3 { get; set; }
    public _A4 Arg4 { get; set; }
    public _A5 Arg5 { get; set; }
    public _A6 Arg6 { get; set; }
    public override int GetArgumentCount() { return 6; }
    public override bool IsReturn() { return false; }
    public override string GetCallbackInfo() { return Callback.Method.Name; }
    public override void MakeCall()
    {
        if( Callback != null )
            Callback(Arg1, Arg2, Arg3, Arg4, Arg5, Arg6);
    }
}
```

If I set the callback to an extern call, it fails calling it (MakeCall()) at runtime with  "System.BadImageFormatException: Method has no body".  This is one of the functions it fails on:

```
        [DllImport(_DllPath)]
        public static extern void wgpu_buffer_map_async(WGpuBuffer buffer, WGpuBufferMapCallbackWrapper callback, IntPtr userData, WGPU_MAP_MODE_FLAGS mode, double offset, double size);
```

And yet it allows this fine:

```
public static void wgpu_buffer_map_async_wrapper(WGpuBuffer buffer, WGpuBufferMapCallbackWrapper callback, IntPtr userData, WGPU_MAP_MODE_FLAGS mode, double offset, double size)
{
    wgpu_buffer_map_async(buffer, callback, userData, mode, offset, size);
}
```
So as long as I dont use the extern function as the callback, it works.  I have hundreds of functions I would need to make wrappers for.  Is there something I am missing in terms of an attribute or some way to not have to wrap every extern function would be great or is this a current known bug?

This is on .NET 8 with the latest pre release package.
Thanks much!

**Comment by SerratedSharp** [comment]
This is a bit outside my comfort zone, but I think you're probably dealing more with the trimming as a result of the IL Linker, which AFAIK happens before AOT.

I have had some success with applying `[DynamicDependency(DynamicallyAccessedMemberTypes.All, typeof(TestsContainer))]` to methods.  The way I think about it is the trimmer traverses code from the root of your application, identifying potentially reachable code.  Anything not reachable from the root of your C# app may get trimmed.  This means methods/callbacks which are only accessed from client side often get trimmed.

The idea is you put it on code that generates the dependency.  I.e. if I have a method `Orchestrator.CreateTests()` that uses reflection to access TestsContainer, then I tag DynamicDependency on CreateTests, and that tells the linker that if it keeps Orchestrator.CreatTests, then it should also keep TestsContainer methods.

I tend to begin by applying `[DynamicDependency(DynamicallyAccessedMemberTypes.All, typeof(ClassOfThingsToKeep))]` on a method that is guaranteed accessible.  I usually start by applying it to Program.Main, which is a heavy handed approach but is good for troubleshooting to begin with.

There's also configuration you can use for the IL Linker:
https://github.com/unoplatform/Uno.Wasm.Bootstrap/blob/main/doc/linker-configuration.md

For example, this is the LinkerConfig.xml I use to make Newtonsoft work with the trimmer:
```
<?xml version="1.0" encoding="UTF-8" ?>
<linker>
	<assembly fullname="System.Core">
		<!-- This is required by JSon.NET and any expression.Compile caller -->
		<type fullname="System.Linq.Expressions*" />
	</assembly>
	<assembly fullname="Newtonsoft.Json" />
</linker>
```

**Comment by BradChase2011** [comment]
Yea I have tried every which way to get it to work by altering the linker descriptor but I cant find any combination of stuff that will get it to work.  Whats weird is I can call the extern function just fine directly.  It seems to only happen when I use it in conjunction with the templated classes.  Unfortunately most of the stuff is WebGPU calls which are not multi threaded so I have to marshal them over using the templates.  I have also created a helper class which calls the functions at the startup to make sure the linker doesnt get rid of them but that didnt change anything either.  I find it odd that it can be called directly but NOT from a templated class... I went ahead and tried the DynamicDependency and it didnt help either unfortunately.  We run 99% of the same code for our other targets, Vulkan, DirectX12, droid, on and on and those all AOT just fine after changing the linker directives and such.  So we know all the pitfalls there generally for what classes cause issues.  For the WASM side we are passed all that except for this interop code which of course doesnt exist in the other platforms and such.  Of all the projects and AOTing, Ive never seen anything like this.

**Comment by jeromelaban** [comment]
Interesting behavior. While I do not have suggestions on how to troubleshoot this, nor would be able to provide an update to the net8 runtime, you should know that the uno bootstrap package is moving to use the official .NET 9 workloads very soon, which means that this kind of discussion should go into the https://github.com/dotnet/runtime repository.

**Comment by BradChase2011** [comment]
I saw you had done some work recently on that and that sounds like good news.  As for us, for now we wrapped them and have AOT working.  Not the best solution but given that it may be temporary anyways sounds good.  This is off topic but with .NET 9, will this essentially be the equivalent of Blazor's current .NET 9 wasm?  Or will the properties in Uno be gone and use the same ones that .net 9 blazor wasm uses?  We have a temp .NET 9 blazor project that we moved away from because of multithreading not being available that we can revive and get working in prep.
